src/Microsoft.AspNet.Mvc.Core/Formatters/AcceptHeaderParser.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Globalization;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNet.Mvc.Formatters
+{
+    public static class AcceptHeaderParser
+    {
+        private static readonly StringSegment QualityParameter = new StringSegment("q");
+
+        public static IList<MediaTypeSegmentWithQuality> ParseAcceptHeader(IList<string> acceptHeaders)
+        {
+            var parsedValues = new List<MediaTypeSegmentWithQuality>();
+            ParseAcceptHeader(acceptHeaders, parsedValues);
+
+            return parsedValues;
+        }
+
+        public static void ParseAcceptHeader(IList<string> acceptHeaders, IList<MediaTypeSegmentWithQuality> parsedValues)
+        {
+            if (acceptHeaders == null)
+            {
+                throw new ArgumentNullException(nameof(acceptHeaders));
+            }
+
+            if (parsedValues == null)
+            {
+                throw new ArgumentNullException(nameof(parsedValues));
+            }
+
+            // If a parser returns an empty list, it means there was no value, but that's valid (e.g. "Accept: ").
+            // The caller can ignore the value.
+            if (acceptHeaders == null)
+            {
+                return;
+            }
+
+            foreach (var value in acceptHeaders)
+            {
+                int index = 0;
+
+                while (!string.IsNullOrEmpty(value) && index < value.Length)
+                {
+                    MediaTypeSegmentWithQuality output;
+                    if (TryParseValue(value, ref index, out output))
+                    {
+                        // The entry may not contain an actual value, like " , "
+                        if (output.MediaType.HasValue)
+                        {
+                            parsedValues.Add(output);
+                        }
+                    }
+                    else
+                    {
+                        throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Invalid values '{0}'.",
+                            value.Substring(index)));
```

Comment:
I'm not 100% convinced we want to throw on an invalid accept value... what's the behavior today?